File: scope104/src/main/java/com/petro/scope104/data/db/DataBaseModule.java
package com.petro.scope104.data.db;

import android.content.Context;

import androidx.room.Room;

import com.petro.scope104.data.UserDataSource;
import com.petro.scope104.data.db.dao.CountryDao;
import com.petro.scope104.data.db.dao.UserDao;
import com.petro.scope104.data.qualifier.DataBaseQualifier;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import dagger.hilt.InstallIn;
import dagger.hilt.android.qualifiers.ApplicationContext;
import dagger.hilt.components.SingletonComponent;

@Module
@InstallIn(SingletonComponent.class)
public abstract class DataBaseModule {
    @Provides
    @Singleton
    public static DataBase provideDataBase(@ApplicationContext Context context) {
       return Room.databaseBuilder(context, DataBase.class, "userDB")
               .allowMainThreadQueries()
               .fallbackToDestructiveMigration()
               .build();
    }

    @Provides
    @Singleton
    public static UserDao providesUserDao(DataBase dataBase) {
        return dataBase.userDao();
    }

    @Provides
    @Singleton
    public static CountryDao providesCountryDao(DataBase dataBase) {
        return dataBase.countryDao();
    }

    @Provides
    @DataBaseQualifier
    @Singleton
    public static UserDataSource userDbDataSource(UserDataSourceDbImpl impl) {
        return impl;
    }
}

File: scope104/src/main/kotlin/com/petro/scope104/data/db/DataBaseModule.kt
package com.petro.scope104.data.db

import android.content.Context
import androidx.room.Room.databaseBuilder
import com.petro.scope104.data.UserDataSource
import com.petro.scope104.data.db.dao.CountryDao
import com.petro.scope104.data.db.dao.UserDao
import com.petro.scope104.data.qualifier.DataBaseQualifier
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.qualifiers.ApplicationContext
import dagger.hilt.components.SingletonComponent
import javax.inject.Singleton

@Module
@InstallIn(SingletonComponent::class)
object DataBaseModule {
    @JvmStatic
    @Provides
    @Singleton
    fun provideDataBase(@ApplicationContext context: Context?): DataBase {
        return databaseBuilder(context!!, DataBase::class.java, "userDB")
            .fallbackToDestructiveMigration()
            .build()
    }

    @JvmStatic
    @Provides
    @Singleton
    fun providesUserDao(dataBase: DataBase): UserDao {
        return dataBase.userDao()
    }

    @JvmStatic
    @Provides
    @Singleton
    fun providesCountryDao(dataBase: DataBase): CountryDao {
        return dataBase.countryDao()
    }

    @JvmStatic
    @Provides
    @DataBaseQualifier
    @Singleton
    fun userDbDataSource(impl: UserDataSourceDbImpl): UserDataSource {
        return impl
    }
}